File: src/main/java/nl/mvdr/adventofcode/adventofcode2024/day08/Part1.java
package nl.mvdr.adventofcode.adventofcode2024.day08;

import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.LongSolver;

public class Part1 implements LongSolver {

    private static final Logger LOGGER = LoggerFactory.getLogger(Part1.class);

    @Override
    public long solve(Stream<String> lines) {
        return CityMap.parse(lines.toList()).countAntinodes();
    }

    public static void main(String[] args) {
        var instance = new Part1();

        var result = instance.solve("input-day08-2024.txt");

        LOGGER.info(result);
    }
}
 

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day08/Part1.kt
package nl.mvdr.adventofcode.adventofcode2025.day08

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.point.Point3D
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart1(lines: Sequence<String>, numberOfPairs: Int = 1000): Int {
    val junctionBoxes = lines.map(Point3D::parse).toList()

    val pairs = findPairs(junctionBoxes, numberOfPairs)

    val circuits = findCircuits(junctionBoxes, pairs)

    return circuits.map { it.size }
        .sorted()
        .takeLast(3)
        .reduce(Int::times)
}

private fun findPairs(junctionBoxes: List<Point3D>, numberOfPairs: Int) =
    junctionBoxes.flatMap { firstBox -> junctionBoxes
            .filter { secondBox -> firstBox != secondBox }
            .map { secondBox -> setOf(firstBox, secondBox) } }
        .distinct()
        .sortedBy { it.first().euclideanDistance(it.last()) }
        .take(numberOfPairs)

private fun findCircuits(junctionBoxes: List<Point3D>, pairs: List<Set<Point3D>>): Set<Set<Point3D?>> {
    val circuits = junctionBoxes.map { setOf(it) }.toMutableSet()

    for (pair in pairs) {
        val firstCircuit = circuits.find { it.contains(pair.first()) }!!
        circuits.remove(firstCircuit)

        val secondCircuit = if (firstCircuit.contains(pair.last())) {
            firstCircuit
        } else {
            circuits.find { it.contains(pair.last()) }!!
        }
        circuits.remove(secondCircuit)

        val newCircuit = firstCircuit union secondCircuit
        circuits.add(newCircuit)
    }

    return circuits
}

fun main() {
    val result = FunctionSolver(::solvePart1).solve("input-day08.txt")
    logger.info { result }
}
